Add some "compile" tests for deriving Associations

weiznich opened the pull request:
Any future cases that should be added?

sgrif commented:
I'd like us to test that these are actually generating the code we expect. We should execute the queries and expect to see some data back.

killercup commented:
Thanks!

I fear this will fail CI with `must_use` errors, though. Let's add insert some data and assert the results.

weiznich commented:
> I'd like us to test that these are actually generating the code we expect. We should execute the queries and expect to see some data back.

There are already tests doing this in diesel_test. So I'm not sure if we should add those tests on two location? 

> I fear this will fail CI with `must_use errors`, though. Let's add insert some data and assert the results.

There is an `#![allow(unused_must_use)]` at the top of the file to prevent this errors.

killercup commented:
> There is an #![allow(unused_must_use)] at the top of the file to prevent this errors.

Lol, I apparently can't read. …And I've just been drinking tea today before anyone asks! 😅

sgrif commented:
The only test I can see which executes a query involving a `foreign_key` option is the test for self referencing associations.

sgrif commented:
We need to provide *some* tests for this behavior, even if it's not testing the actual results of executing the query. I'd like to at least test that `foreign_key` is doing the right thing, and probably a test like:

```rust
assert_eq!(
    debug_sql(Foo::belonging_to(&bar)),
    debug_sql(foos.filter(bar_id.eq(bar.id))),
);
```

weiznich commented:
>  I'd like to at least test that foreign_key is doing the right thing, and probably a test like:

I've adjusted the diesel_codegen tests to check the generated sql

> The only test I can see which executes a query involving a foreign_key option is the test for self referencing associations.

I've added one test to diesel_tests for this scenario.

killercup commented on diesel_codegen/tests/associations.rs:
```diff
@@ -0,0 +1,149 @@
+        .select((users::id, users::name))
+        .filter(posts::id.eq(1)
+            .and(posts::user_id.eq(2))
+                .and(posts::title.eq("Bar")));
```
Can you name these "compile-success assertions" by making them let statements? E.g., `let _can_select_columns_from_join = posts::table.inn…` (or a better name)

killercup commented on diesel_codegen/tests/associations.rs:
```diff
@@ -0,0 +1,149 @@
+        id: i32,
+        parent_id: Option<i32>,
+    }
+    let t = Tree{ id: 42, parent_id: None };
```
pls add space before `{` (sorry my brain zooms in on inconsistencies like that)

killercup commented on diesel_tests/tests/associations.rs:
```diff
@@ -187,3 +187,70 @@ fn conn_with_test_data() -> (TestConnection, User, User, User) {
+        id: i32,
+        belongs_to_user: i32,
+        title: String,
+    }
```
Please add empty lines before each of the above items where missing

killercup commented on diesel_tests/tests/associations.rs:
```diff
@@ -187,3 +187,70 @@ fn conn_with_test_data() -> (TestConnection, User, User, User) {
+        }
+    }
+
+    #[derive(Debug, Queryable, Identifiable, PartialEq, Clone)]
```
Can you put diesel's derives last? Or… sort them alphabetically? Or anything that I can identify as 'order'? Or reveal the system you ordered them by? :D Sorry for commenting on all the nits

killercup commented on diesel_tests/tests/associations.rs:
```diff
@@ -187,3 +187,70 @@ fn conn_with_test_data() -> (TestConnection, User, User, User) {
+    let post2 = Post { id: 2, belongs_to_user: 2, title: "World".into() };
+    let post3 = Post { id: 3, belongs_to_user: 1, title: "Hello 2".into() };
+
+
```
✂️

killercup commented on diesel_codegen/tests/associations.rs:
```diff
@@ -0,0 +1,149 @@
+    assert_eq!(
+        debug_sql::<Backend, _>(&belong_to),
+        debug_sql::<Backend, _>(&filter)
+    )
```
The macro may return `()`, but in other tests we always append a `;`

weiznich commented:
Any idea why the tests are failing? It builds fine for me locally using the same rust version (1.19.0)
